File: server/http/responses/game.go
package responses

import (
	"app/models"
	"time"
)

type GameResource struct {
	Code          string            `json:"code"`
	Bet           uint              `json:"bet"`
	WinningPoints uint              `json:"winning_points"`
	JoinType      string            `json:"join_type"`
	StartedAt     *time.Time        `json:"started_at"`
	Currency      *CurrencyResource `json:"currency"`
	Players       []PlayerResource  `json:"players"`
	Creator       *UserResource     `json:"creator"`
	PlayersCount  uint              `json:"players_count"`
}

type PlayerResource struct {
	ID       uint   `json:"id"`
	Username string `json:"username"`
	IsHost   bool   `json:"is_host"`
}

func NewPlayerResource(user models.User, isHost bool) PlayerResource {
	return PlayerResource{
		ID:       user.ID,
		Username: user.Username,
		IsHost:   isHost,
	}
}

func NewGameResource(game models.Game) GameResource {
	rs := GameResource{
		StartedAt:     game.StartedAt,
		Code:          game.Code,
		Bet:           game.Bet,
		WinningPoints: game.WinningPoints,
		JoinType:      game.JoinType,
		PlayersCount:  game.PlayersCount,
	}
	if game.Creator.ID != 0 {
		creator := NewUserResource(game.Creator)
		rs.Creator = &creator
	}

	if game.Currency.ID != 0 {
		c := NewCurrencyResource(game.Currency)
		rs.Currency = &c
	}

	if game.Players != nil {
		for _, p := range game.Players {
			rs.Players = append(rs.Players, NewPlayerResource(p, p.ID == game.CreatorID))
		}
	}

	return rs
}

File: client/src/api/game.ts
import fetchApi from "@/packages/fetchApi.ts";
import type {Currency} from "@/api/currency.ts";
import type {User} from "@/api/user.ts";

export enum JoinType {
  ANYONE = 'anyone',
  FRIENDS = 'friends',
  LINK = 'link'
}

export interface CreateGameInput {
  currency_id: number
  bet: number
  winning_points: number
  join_type: JoinType
}

export interface Game {
  code: string
  bet: number
  creator_id: number
  winning_points: number
  join_type: JoinType
  currency?: Currency
  creator?: User
  players: [Player]
  players_count: number
}

export interface Player {
  id: number
  username: string
  is_host: boolean
}

export const createGame = async (input: CreateGameInput): Promise<Game> => {
  const data = await fetchApi.post('/games', input)

  return data.data;
}

export const getLastCreatedGame = async (): Promise<Game | null> => {
  const {data} = await fetchApi.get('/games/current');

  return data
}

export const getGame = async (code: string): Promise<Game | null> => {
  const {data} = await fetchApi.get(`/games/${code}`);

  return data
}

export const leaveGame = async () => {
  return fetchApi.delete(`/games/leave`)
}

export interface GetGamesInput {
  page: number
  search?: string
}

export const getGames = async (params: GetGamesInput) => {
  const {data} = await fetchApi.get('/games', {
    params: params
  });

  return data;
}

export const joinGame = async (code: string) => {
  return fetchApi.post(`/games/join/${code}`)
}
